aws/resource_aws_dms_endpoint.go

```diff
@@ -333,13 +409,41 @@ func resourceAwsDmsEndpointSetState(d *schema.ResourceData, endpoint *dms.Endpoi
 	d.Set("endpoint_type", strings.ToLower(*endpoint.EndpointType))
 	d.Set("engine_name", endpoint.EngineName)
 
-	if *endpoint.EngineName == "dynamodb" {
+	switch *endpoint.EngineName {
+	case "dynamodb":
 		if endpoint.DynamoDbSettings != nil {
 			d.Set("service_access_role", endpoint.DynamoDbSettings.ServiceAccessRoleArn)
 		} else {
 			d.Set("service_access_role", "")
 		}
-	} else {
+	case "mongodb":
+		if endpoint.MongoDbSettings != nil {
+			d.Set("username", endpoint.MongoDbSettings.Username)
+			d.Set("server_name", endpoint.MongoDbSettings.ServerName)
+			d.Set("port", endpoint.MongoDbSettings.Port)
+			d.Set("database_name", endpoint.MongoDbSettings.DatabaseName)
+
+			d.Set("mongodb_settings.0.auth_type", endpoint.MongoDbSettings.AuthType)
```

Comment:
To set `TypeList` or `TypeSet` nested attributes into the Terraform state, we need to create a function that outputs a slice. These current `d.Set()` calls are likely returning an error that is not caught and silently ignoring the issue, which means that Terraform is going to "pass through" the Terraform configuration as-is into the state and it will never get updated from the API.

Here's an example implementation:

```go
// Always perform error checking when setting non-scalar (TypeList, TypeSet, TypeMap) attributes
if err := d.Set("mongo_settings", flattenDmsMongoDbSettings(endpoint.MongoDbSettings)); err != nil {
  return fmt.Errorf("error setting mongo_settings: %s", err)
}

// Outside the read function
func flattenDmsMongoDbSettings(settings *dms.MongoDbSettings) []map[string]interface{} {
  if settings == nil {
    return []map[string]interface{}{}
  }

  m := map[string]interface{}{
    "auth_mechanism": aws.StringValue(settings.AuthMechanism),
    "auth_source": aws.StringValue(settings.AuthSource),
    "docs_to_investigate": aws.StringValue(settings.DocsToInvestigate),
    "extract_doc_id": aws.StringValue(settings.ExtractDocId),
    "nesting_level": aws.StringValue(settings.NestingLevel),
  }

  return []map[string]interface{}{m}
}
```